.github/workflows/test.yml

```diff
@@ -0,0 +1,71 @@
+name: test
+
+on: [pull_request]
+
+
+jobs:
+  unit-test:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ['3.7']
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v1
+      - name: Configure Python
+        uses: actions/setup-python@v1
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Apt-Get
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y shellcheck sqlite3 bsd-mailx
+      - name: Install
+        run: |
+          pip install ."[all]"
+          pip install pytest-xdist
+      - name: Style
```

Comment:
Perhaps I'm being fussy, but is there any way to refactor all the set-up above this point since it gets re-used every time?

Reply:
I did try to avoid this but:

* The setup for the two jobs is not actually the same.
* GitHub actions does not support yaml anchors.

Reply:
> is there a reason why unit-test is a hyphen and functional_test and underscore?

Because I fluffed up a rebase.

Reply:
> I'm imagining `unit-test: needs: common-setup`

The [GH Actions `needs`](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idneeds) keyword is like a dependency in Cylc, it says a must run before b, but these are different jobs running in different environments, they don't share a container.

What we can do later (see todos) is to do the setup, cache dependencies then load that setup in future jobs. BUT this is easier said than done due to our need to install stuff via `apt-get`. Caching Python modules is easy enough, caching system packages is harder. I did some digging a while back and found a stackoverflow post where someone showed how to cache a particular dependency installed via apt-get, but there isn't a generic solution. Note GitHub has not provided any actions for working with apt-get.

Reply:
None of this is really blocking - just me trying to gain a better understanding.

Reply:
No, it's a good point and one which irritates me!

Reply:
I've managed to consolidate things down to something pretty small now anyhoo.